![](https://github.com/VascoCapao/images/blob/main/Association%20Rules.png?raw=true)

In [22]:
import pandas as pd

from AA_functions import association_rules_pipeline,dataframe_summary


In [23]:
customer_basket = pd.read_csv('customer_basket.csv')

Diversifiers = pd.read_csv('custom_cluster_csv_files\\Cluster Diversifiers.csv', index_col = 0)
Occasional_discount_shoppers = pd.read_csv('custom_cluster_csv_files\\Cluster Occasional discount shoppers.csv', index_col = 0)
Large_Families = pd.read_csv('custom_cluster_csv_files\\Cluster Large Families.csv', index_col = 0)
Vegetarians = pd.read_csv('custom_cluster_csv_files\\Cluster Vegetarians.csv', index_col = 0)
Pet_Owners = pd.read_csv('custom_cluster_csv_files\\Cluster Pet Owners.csv', index_col = 0)
Young_Shoppers = pd.read_csv('custom_cluster_csv_files\\Cluster Young Shoppers.csv', index_col = 0)
Premium_Clients = pd.read_csv('custom_cluster_csv_files\\Cluster Premium Clients.csv', index_col = 0)

In [24]:
dataframe_summary(customer_basket)

-------------------------------------------------------------------------------------------------------------------------------------------------
SHAPE of Dataset -----------------------------------------------------------------------------------------------------------------------
(89952, 3)
-------------------------------------------------------------------------------------------------------------------------------------------------
DATA TYPES of Dataset ------------------------------------------------------------------------------------------------------------------
invoice_id        int64
list_of_goods    object
customer_id       int64
dtype: object
-------------------------------------------------------------------------------------------------------------------------------------------------
MISSING VALUES in Dataset --------------------------------------------------------------------------------------------------------------
invoice_id       0
list_of_goods    0
customer_id     

In [25]:
clusters = {'Diversifiers': Diversifiers, 
            'Occasional Discount Shoppers': Occasional_discount_shoppers, 
            'Large_Families': Large_Families, 
            'Vegetarians': Vegetarians, 
            'Pet_Owners': Pet_Owners, 
            'Young_Shoppers': Young_Shoppers, 
            'Premium_Clients': Premium_Clients 
            }

for cluster_name, cluster in clusters.items():
    print(cluster_name.upper())
    display(association_rules_pipeline(cluster, customer_basket).sort_values('lift', ascending = False)[:60])

DIVERSIFIERS


antecedents              consequents  antecedent support  \
17                 ('cake')   ('cooking oil', 'oil')            0.439530   
16   ('cooking oil', 'oil')                 ('cake')            0.474914   
18          ('cooking oil')          ('cake', 'oil')            0.569789   
15          ('cake', 'oil')          ('cooking oil')            0.368357   
19                  ('oil')  ('cake', 'cooking oil')            0.796948   
14  ('cake', 'cooking oil')                  ('oil')            0.266797   
13              ('muffins')                  ('oil')            0.250523   
12                  ('oil')              ('muffins')            0.796948   
1           ('cooking oil')                 ('cake')            0.569789   
0                  ('cake')          ('cooking oil')            0.439530   
5            ('candy bars')          ('cooking oil')            0.349517   
4           ('cooking oil')           ('candy bars')            0.569789   
3                   ('oil')                 ('cake')            0.796948   
2                  ('cake')                  ('oil')            0.439530   
9                   ('oil')          ('cooking oil')            0.796948   
8           ('cooking oil')                  ('oil')            0.569789   
11                  ('oil')                 ('gums')            0.796948   
10                 ('gums')                  ('oil')            0.303194   
7            ('candy bars')                  ('oil')            0.349517   
6                   ('oil')           ('candy bars')            0.796948   

    consequent support   support  confidence      lift  leverage  conviction  \
17            0.474914  0.230130    0.523583  1.102479  0.021391    1.102156   
16            0.439530  0.230130    0.484573  1.102479  0.021391    1.087389   
18            0.368357  0.230130    0.403887  1.096456  0.020245    1.059603   
15            0.569789  0.230130    0.624748  1.096456  0.020245    1.146460   
19            0.266797  0.230130    0.288765  1.082337  0.017507    1.030886   
14            0.796948  0.230130    0.862566  1.082337  0.017507    1.477457   
13            0.796948  0.212776    0.849326  1.065724  0.013122    1.347627   
12            0.250523  0.212776    0.266989  1.065724  0.013122    1.022463   
1             0.439530  0.266797    0.468239  1.065317  0.016358    1.053988   
0             0.569789  0.266797    0.607006  1.065317  0.016358    1.094701   
5             0.569789  0.211696    0.605680  1.062991  0.012545    1.091021   
4             0.349517  0.211696    0.371534  1.062991  0.012545    1.035032   
3             0.439530  0.368357    0.462210  1.051600  0.018075    1.042172   
2             0.796948  0.368357    0.838070  1.051600  0.018075    1.253953   
9             0.569789  0.474914    0.595916  1.045854  0.020822    1.064658   
8             0.796948  0.474914    0.833491  1.045854  0.020822    1.219469   
11            0.303194  0.252684    0.317065  1.045749  0.011054    1.020311   
10            0.796948  0.252684    0.833408  1.045749  0.011054    1.218856   
7             0.796948  0.291174    0.833076  1.045333  0.012627    1.216433   
6             0.349517  0.291174    0.365362  1.045333  0.012627    1.024966   

    zhangs_metric  
17       0.165849  
16       0.177025  
18       0.204482  
15       0.139272  
19       0.374651  
14       0.103755  
13       0.082285  
12       0.303717  
1        0.142517  
0        0.109395  
5        0.091099  
4        0.137742  
3        0.241653  
2        0.087548  
9        0.215924  
8        0.101913  
11       0.215451  
10       0.062783  
7        0.066669  
6        0.213575

OCCASIONAL DISCOUNT SHOPPERS


antecedents      consequents  antecedent support  consequent support  \
5          ('oil')  ('cooking oil')            0.553174            0.408015   
4  ('cooking oil')          ('oil')            0.408015            0.553174   
3   ('candy bars')          ('oil')            0.259608            0.553174   
2          ('oil')   ('candy bars')            0.553174            0.259608   
0         ('cake')          ('oil')            0.319975            0.553174   
1          ('oil')         ('cake')            0.553174            0.319975   

    support  confidence      lift  leverage  conviction  zhangs_metric  
5  0.320692    0.579730  1.420854  0.094988    1.408582       0.662894  
4  0.320692    0.785980  1.420854  0.094988    2.087776       0.500347  
3  0.202911    0.781605  1.412946  0.059302    2.045956       0.394735  
2  0.202911    0.366812  1.412946  0.059302    1.169308       0.654078  
0  0.249483    0.779697  1.409495  0.072481    2.028230       0.427228  
1  0.249483    0.451003  1.409495  0.072481    1.238668       0.650200

LARGE_FAMILIES


antecedents                     consequents  \
17                        ('cake')  ('babies food', 'cooking oil')   
16  ('babies food', 'cooking oil')                        ('cake')   
18                 ('cooking oil')         ('cake', 'babies food')   
15         ('cake', 'babies food')                 ('cooking oil')   
19                 ('babies food')         ('cake', 'cooking oil')   
14         ('cake', 'cooking oil')                 ('babies food')   
11                 ('cooking oil')                        ('cake')   
10                        ('cake')                 ('cooking oil')   
12                 ('cooking oil')                  ('candy bars')   
13                  ('candy bars')                 ('cooking oil')   
8                  ('babies food')                     ('muffins')   
9                      ('muffins')                 ('babies food')   
3                   ('candy bars')                 ('babies food')   
2                  ('babies food')                  ('candy bars')   
1                  ('babies food')                        ('cake')   
0                         ('cake')                 ('babies food')   
7                  ('babies food')                        ('gums')   
6                         ('gums')                 ('babies food')   
5                  ('cooking oil')                 ('babies food')   
4                  ('babies food')                 ('cooking oil')   

    antecedent support  consequent support   support  confidence      lift  \
17            0.438785            0.473341  0.229772    0.523656  1.106296   
16            0.473341            0.438785  0.229772    0.485426  1.106296   
18            0.574689            0.364129  0.229772    0.399821  1.098020   
15            0.364129            0.574689  0.229772    0.631020  1.098020   
19            0.792599            0.270066  0.229772    0.289898  1.073430   
14            0.270066            0.792599  0.229772    0.850800  1.073430   
11            0.574689            0.438785  0.270066    0.469935  1.070990   
10            0.438785            0.574689  0.270066    0.615486  1.070990   
12            0.574689            0.362452  0.220489    0.383666  1.058528   
13            0.362452            0.574689  0.220489    0.608324  1.058528   
8             0.792599            0.264071  0.220102    0.277696  1.051599   
9             0.264071            0.792599  0.220102    0.833496  1.051599   
3             0.362452            0.792599  0.300883    0.830132  1.047354   
2             0.792599            0.362452  0.300883    0.379616  1.047354   
1             0.792599            0.438785  0.364129    0.459411  1.047006   
0             0.438785            0.792599  0.364129    0.829856  1.047006   
7             0.792599            0.303204  0.251499    0.317309  1.046520   
6             0.303204            0.792599  0.251499    0.829471  1.046520   
5             0.574689            0.792599  0.473341    0.823648  1.039174   
4             0.792599            0.574689  0.473341    0.597202  1.039174   

    leverage  conviction  zhangs_metric  
17  0.022077    1.105625       0.171204  
16  0.022077    1.090640       0.182438  
18  0.020512    1.059469       0.209893  
15  0.020512    1.152666       0.140389  
19  0.015718    1.027927       0.329831  
14  0.015718    1.390086       0.093717  
11  0.017901    1.058766       0.155850  
10  0.017901    1.106101       0.118110  
12  0.012191    1.034419       0.130004  
13  0.012191    1.085876       0.086726  
8   0.010800    1.018864       0.236581  
9   0.010800    1.245624       0.066674  
3   0.013604    1.220952       0.070917  
2   0.013604    1.027666       0.217998  
1   0.016348    1.038154       0.216469  
0   0.016348    1.218975       0.079998  
7   0.011180    1.020661       0.214329  
6   0.011180    1.216219       0.063795  
5   0.017844    1.176065       0.088635  
4   0.017844    1.055891       0.181761

VEGETARIANS


antecedents                     consequents  \
21       ('tomatoes', 'asparagus')               ('mashed potato')   
24               ('mashed potato')       ('tomatoes', 'asparagus')   
25                   ('asparagus')   ('tomatoes', 'mashed potato')   
20   ('tomatoes', 'mashed potato')                   ('asparagus')   
22  ('mashed potato', 'asparagus')                    ('tomatoes')   
23                    ('tomatoes')  ('mashed potato', 'asparagus')   
19                     ('carrots')       ('tomatoes', 'asparagus')   
14       ('tomatoes', 'asparagus')                     ('carrots')   
18                   ('asparagus')         ('tomatoes', 'carrots')   
15         ('tomatoes', 'carrots')                   ('asparagus')   
16        ('asparagus', 'carrots')                    ('tomatoes')   
17                    ('tomatoes')        ('asparagus', 'carrots')   
6                ('mashed potato')                     ('carrots')   
7                      ('carrots')               ('mashed potato')   
2                ('mashed potato')                   ('asparagus')   
3                    ('asparagus')               ('mashed potato')   
13                      ('melons')                    ('tomatoes')   
12                    ('tomatoes')                      ('melons')   
11               ('mashed potato')                    ('tomatoes')   
10                    ('tomatoes')               ('mashed potato')   
1                      ('carrots')                   ('asparagus')   
0                    ('asparagus')                     ('carrots')   
9                      ('carrots')                    ('tomatoes')   
8                     ('tomatoes')                     ('carrots')   
5                    ('asparagus')                    ('tomatoes')   
4                     ('tomatoes')                   ('asparagus')   

    antecedent support  consequent support   support  confidence      lift  \
21            0.547748            0.348069  0.210811    0.384868  1.105723   
24            0.348069            0.547748  0.210811    0.605657  1.105723   
25            0.670142            0.287194  0.210811    0.314577  1.095344   
20            0.287194            0.670142  0.210811    0.734035  1.095344   
22            0.247683            0.784041  0.210811    0.851130  1.085568   
23            0.784041            0.247683  0.210811    0.268877  1.085568   
19            0.559846            0.547748  0.332432    0.593793  1.084063   
14            0.547748            0.559846  0.332432    0.606908  1.084063   
18            0.670142            0.458752  0.332432    0.496063  1.081332   
15            0.458752            0.670142  0.332432    0.724646  1.081332   
16            0.393372            0.784041  0.332432    0.845084  1.077857   
17            0.784041            0.393372  0.332432    0.423999  1.077857   
6             0.348069            0.559846  0.209138    0.600850  1.073243   
7             0.559846            0.348069  0.209138    0.373563  1.073243   
2             0.348069            0.670142  0.247683    0.711592  1.061853   
3             0.670142            0.348069  0.247683    0.369599  1.061853   
13            0.247619            0.784041  0.204569    0.826143  1.053699   
12            0.784041            0.247619  0.204569    0.260916  1.053699   
11            0.348069            0.784041  0.287194    0.825106  1.052376   
10            0.784041            0.348069  0.287194    0.366300  1.052376   
1             0.559846            0.670142  0.393372    0.702644  1.048500   
0             0.670142            0.559846  0.393372    0.586998  1.048500   
9             0.559846            0.784041  0.458752    0.819425  1.045130   
8             0.784041            0.559846  0.458752    0.585112  1.045130   
5             0.670142            0.784041  0.547748    0.817361  1.042498   
4             0.784041            0.670142  0.547748    0.698621  1.042498   

    leverage  conviction 

PET_OWNERS


antecedents                      consequents  \
34                  ('canned_tuna')        ('fresh tuna', 'catfish')   
31        ('fresh tuna', 'catfish')                  ('canned_tuna')   
39                ('grated cheese')         ('shrimp', 'shower gel')   
38         ('shrimp', 'shower gel')                ('grated cheese')   
0                   ('canned_tuna')                      ('catfish')   
16                ('grated cheese')                   ('shower gel')   
1                       ('catfish')                  ('canned_tuna')   
30    ('fresh tuna', 'canned_tuna')                      ('catfish')   
35                      ('catfish')    ('fresh tuna', 'canned_tuna')   
36      ('shrimp', 'grated cheese')                   ('shower gel')   
41                   ('shower gel')      ('shrimp', 'grated cheese')   
17                   ('shower gel')                ('grated cheese')   
33                   ('fresh tuna')       ('canned_tuna', 'catfish')   
32       ('canned_tuna', 'catfish')                   ('fresh tuna')   
3                   ('canned_tuna')                   ('fresh tuna')   
2                    ('fresh tuna')                  ('canned_tuna')   
22                    ('olive oil')                   ('shower gel')   
45                       ('salmon')          ('shrimp', 'olive oil')   
47                    ('olive oil')             ('salmon', 'shrimp')   
10                       ('salmon')                  ('cooking oil')   
8                    ('fresh tuna')                  ('cooking oil')   
4                    ('fresh tuna')                      ('catfish')   
11                  ('cooking oil')                       ('salmon')   
44          ('shrimp', 'olive oil')                       ('salmon')   
5                       ('catfish')                   ('fresh tuna')   
42             ('salmon', 'shrimp')                    ('olive oil')   
9                   ('cooking oil')                   ('fresh tuna')   
23                   ('shower gel')                    ('olive oil')   
15                    ('olive oil')                   ('fresh tuna')   
14                   ('fresh tuna')                    ('olive oil')   
20                       ('salmon')                    ('olive oil')   
21                    ('olive oil')                       ('salmon')   
46                       ('shrimp')          ('salmon', 'olive oil')   
40                       ('shrimp')  ('shower gel', 'grated cheese')   
37  ('shower gel', 'grated cheese')                       ('shrimp')   
43          ('salmon', 'olive oil')                       ('shrimp')   
19                ('grated cheese')                       ('shrimp')   
18                       ('shrimp')                ('grated cheese')   
27                       ('shrimp')                       ('salmon')   
26                       ('salmon')                       ('shrimp')   
25                    ('olive oil')                       ('shrimp')   
24                       ('shrimp')                    ('olive oil')   
12                  ('cooking oil')                       ('shrimp')   
7                       ('catfish')                       ('shrimp')   
6                        ('shrimp')                      ('catfish')   
13                       ('shrimp')                  ('cooking oil')   
29                   ('shower gel')                       ('shrimp')   
28                       ('shrimp')                   ('shower gel')   

    antecedent support  consequent support   support  confidence      lift  \
34            0.250000            0.250000  0.250000    1.000000  4.000000   
31            0.250000            0.250000  0.250000    1.000000  4.000000   
39            0.250000            0.250000  0.250000    1.000000  4.000000   
38            0.250000            0.250000  0.250000    1.000000  4.000000   
0             0.250000            0.333333  0.250000    1.000000  3.000000   
16            0.250000     

YOUNG_SHOPPERS


antecedents       consequents  antecedent support  consequent support  \
1          ('beer')    ('white wine')            0.295639            0.711013   
0    ('white wine')          ('beer')            0.711013            0.295639   
5    ('white wine')  ('dessert wine')            0.711013            0.338507   
4  ('dessert wine')    ('white wine')            0.338507            0.711013   
2         ('cider')    ('white wine')            0.488544            0.711013   
3    ('white wine')         ('cider')            0.711013            0.488544   

    support  confidence      lift  leverage  conviction  zhangs_metric  
1  0.231338    0.782500  1.100543  0.021134    1.328678       0.129703  
0  0.231338    0.325364  1.100543  0.021134    1.044060       0.316131  
5  0.264597    0.372141  1.099361  0.023914    1.053570       0.312749  
4  0.264597    0.781659  1.099361  0.023914    1.323562       0.136631  
2  0.372506    0.762481  1.072388  0.025145    1.216692       0.131979  
3  0.372506    0.523909  1.072388  0.025145    1.074281       0.233579

PREMIUM_CLIENTS


antecedents               consequents  antecedent support  \
6              ('champagne')                ('laptop')            0.738149   
7                 ('laptop')             ('champagne')            0.257301   
4      ('samsung galaxy 10')  ('bluetooth headphones')            0.507688   
5   ('bluetooth headphones')     ('samsung galaxy 10')            0.387408   
3   ('bluetooth headphones')             ('champagne')            0.387408   
2              ('champagne')  ('bluetooth headphones')            0.738149   
10             ('champagne')             ('spaghetti')            0.738149   
11             ('spaghetti')             ('champagne')            0.300155   
9      ('samsung galaxy 10')             ('champagne')            0.507688   
8              ('champagne')     ('samsung galaxy 10')            0.738149   
0              ('champagne')               ('airpods')            0.738149   
1                ('airpods')             ('champagne')            0.349286   

    consequent support   support  confidence      lift  leverage  conviction  \
6             0.257301  0.203075    0.275114  1.069228  0.013148    1.024573   
7             0.738149  0.203075    0.789250  1.069228  0.013148    1.242472   
4             0.387408  0.209626    0.412903  1.065810  0.012944    1.043426   
5             0.507688  0.209626    0.541099  1.065810  0.012944    1.072807   
3             0.738149  0.303521    0.783466  1.061393  0.017556    1.209284   
2             0.387408  0.303521    0.411192  1.061393  0.017556    1.040394   
10            0.300155  0.235010    0.318378  1.060713  0.013451    1.026735   
11            0.738149  0.235010    0.782965  1.060713  0.013451    1.206488   
9             0.738149  0.395960    0.779928  1.056600  0.021211    1.189842   
8             0.507688  0.395960    0.536423  1.056600  0.021211    1.061985   
0             0.349286  0.272132    0.368668  1.055490  0.014307    1.030700   
1             0.738149  0.272132    0.779109  1.055490  0.014307    1.185430   

    zhangs_metric  
6        0.247264  
7        0.087177  
4        0.125422  
5        0.100796  
3        0.094421  
2        0.220896  
10       0.218589  
11       0.081786  
9        0.108808  
8        0.204573  
0        0.200773  
1        0.080792

## Diversifiers

The association analysis results indicate that the products 'cake', 'cooking oil', and 'oil' are frequently purchased together. When 'cake' is acquired, there's a tendency for the simultaneous purchase of 'cooking oil' and 'oil', and vice versa. However, the associations between these products are moderate, with confidences around 40-50% and lifts of approximately 1.10. While these associations suggest some correlation between the items, their effects on purchase decisions may not be highly significant, as confidences and lifts are relatively modest. 

## Occasionals discount shoppers

The analysis of occasional discount shoppers reveals patterns in purchasing behavior, particularly regarding the items 'oil', 'cooking oil', and 'candy bars'. These shoppers often show a preference for specific combinations of these products, such as 'oil' leading to the purchase of 'cooking oil' or 'candy bars'. Confidence levels are relatively high, ranging from 57.97% to 78.60%, indicating a strong likelihood of these associations occurring. The lift values, which range around 1.41 to 1.42, suggest that these associations are more frequent than random chance.


## Large Families

The association analysis reveals significant patterns in purchasing behavior, particularly concerning the items 'cake', 'cooking oil', and 'babies food'. These items frequently appear together in shopping transactions, indicating potential preferences or needs among large families. For instance, 'cake' is often purchased alongside 'babies food' and 'cooking oil', and vice versa. Confidence levels range from 45.94% to 85.08%, suggesting strong associations between these items. Lift values, which range from 1.04 to 1.10, indicate that these associations are more likely than random chance. However, while statistically significant, the practical impact of these associations on consumer behavior may vary. 

## Vegetarians
The association analysis reveals purchasing patterns among items commonly consumed by vegetarians, such as 'tomatoes', 'asparagus', 'mashed potato', and 'carrots'. There are frequent associations between these items within transactions, indicating potential dietary preferences or meal compositions among vegetarians. Confidence levels range between 36.93% and 85.11%, suggesting moderate to strong associations between these items. Lift values, ranging from 1.04 to 1.11, further support the significance of these associations.

## Pet Owners
Notably, there are strong associations between certain types of fish products such as 'canned tuna', 'fresh tuna', 'catfish', and other related items. These associations are characterized by high confidence levels of 100% and lift values of 4, indicating a strong likelihood of purchasing these items together. Additionally, there are associations between 'grated cheese', 'shrimp', 'shower gel', and 'olive oil', suggesting potential purchasing patterns among pet owners.

## Young Shoppers
The association analysis suggests that young shoppers exhibit interesting preferences in alcoholic beverages. Specifically, there are significant associations between 'beer' and 'white wine', as well as between 'white wine' and 'dessert wine'. These associations indicate that young shoppers are likely to purchase beer and white wine together, as well as white wine and dessert wine together. Additionally, there is a notable association between 'cider' and 'white wine', suggesting that young shoppers may also favor cider along with white wine. These associations are characterized by moderate to high confidence levels and lift values, indicating a likelihood of these items being purchased together by young shoppers.

## Premium Clients
For premium clients, analysis reveals interesting trends. There's a moderate link between purchasing champagne and laptops, with around 27% confidence and a slight uplift. Similarly, Samsung Galaxy 10 purchasers tend to buy Bluetooth headphones, and vice versa, with about 41% confidence. Additionally, there's an association between champagne and Bluetooth headphones, spaghetti, and Airpods, with moderate confidence levels. However, these associations may not significantly influence purchase decisions due to low lift values.